### PR TITLE
Enable forkserver as a default forking mode for HPO

### PR DESCRIPTION
**Pull request by gradientsky**
*Description of changes:*
- fixes issues with threads locking
- fixes issues with CUDA - [#620](https://github.com/awslabs/autogluon/issues/602)
- fixes to scheduler tests to enable pickling
- use forkserver only for HPO; dataloader updates
- force thread_pool usage for forkserver to prevent GIL locks
- Added GPU control options; added docs
- Added stdout/stderr capture

Updated version of #767 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-788-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-788/1/index.html

**Comment by szha** [comment]
Job PR-788-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-788/2/index.html

**Comment by szha** [comment]
Job PR-788-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-788/3/index.html

File: tabular/src/autogluon/tabular/models/tabular_nn/tabular_nn_model.py [hunk 1]
```diff
@@ -674,8 +670,9 @@ def load(cls, path: str, reset_paths=True, verbose=True):
         return model
 
     def hyperparameter_tune(self, X_train, y_train, X_val, y_val, scheduler_options, **kwargs):
-        time_start = time.time()
         """ Performs HPO and sets self.params to best hyperparameter values """
+        time_start = time.time()
+        force_forkserver()
```
Comment (sxjscience) [inline tabular/src/autogluon/tabular/models/tabular_nn/tabular_nn_model.py:675]: I noticed that there is no ` from autogluon.core.utils.multiprocessing_utils import force_forkserver `

Comment (gradientsky) [reply tabular/src/autogluon/tabular/models/tabular_nn/tabular_nn_model.py:676]: fixed.

**Comment by szha** [comment]
Job PR-788-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-788/4/index.html

**Comment by mseeger** [comment]
What remains to be done here? Is there still a review required?

**Comment by gradientsky** [comment]
> What remains to be done here? Is there still a review required?
Waiting #733 to merge in to rebase before finalizing.

**Comment by szha** [comment]
Job PR-788-7 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-788/7/index.html

**Comment by gradientsky** [comment]
Updated branch to include recent master. Still waiting for #733 to finalize.

**Comment by szha** [comment]
Job PR-788-9 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-788/9/index.html

**Comment by szha** [comment]
Job PR-788-10 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-788/10/index.html

File: tabular/src/autogluon/tabular/models/fastainn/tabular_nn_fastai.py [hunk 2]
```diff
@@ -139,10 +141,13 @@ def preprocess_train(self, X_train, y_train, X_val, y_val, **kwargs):
...
         label_class = FloatList if self.problem_type == REGRESSION else None
         procs = [FillMissing, Categorify, Normalize]
+
+        # additional workers are helping only when fork is enabled; in other mp modes, communication overhead reduces performance
+        num_workers = defaults.cpus if is_fork_enabled() else 0
         data = (TabularList.from_df(df_train, path=self.path, cat_names=self.cat_columns, cont_names=self.cont_columns, procs=procs)
                 .split_by_idxs(train_idx, val_idx)
                 .label_from_df(cols=LABEL, label_cls=label_class)
-                .databunch(bs=self.params['bs'] if len(X_train) > self.params['bs'] else 32))
+                .databunch(bs=self.params['bs'] if len(X_train) > self.params['bs'] else 32, num_workers=num_workers))
```
Comment (Innixma) [inline tabular/src/autogluon/tabular/models/fastainn/tabular_nn_fastai.py:150]: Is this a speedup from mainline when we aren't doing HPO?

**Review by Innixma — APPROVED** [review]
Looks good!

**Review by sxjscience — APPROVED** [review]

File: tabular/src/autogluon/tabular/models/fastainn/tabular_nn_fastai.py (hunk 2 above)
Comment (gradientsky) [reply tabular/src/autogluon/tabular/models/fastainn/tabular_nn_fastai.py:150]: It's a speedup for HPO, for non-HPO the performance will be the same as it was before this PR (unchanged)

**Pull request merged**